File: iot_driver_copilot/industrial_test_4/modbus.py
import time
from typing import List

try:
    import serial  # pyserial
except ImportError as e:
    raise ImportError('pyserial is required: pip install pyserial')


class ModbusError(Exception):
    pass


class ModbusRTU:
    def __init__(self, port: str, baudrate: int, bytesize: int, parity: str, stopbits: int, timeout_s: float):
        self.port = port
        self.baudrate = baudrate
        self.bytesize = bytesize
        self.parity = parity
        self.stopbits = stopbits
        self.timeout_s = timeout_s
        self.ser: serial.Serial | None = None

    def open(self):
        if self.ser and self.ser.is_open:
            return
        parity_map = {
            'N': serial.PARITY_NONE,
            'E': serial.PARITY_EVEN,
            'O': serial.PARITY_ODD,
            'M': serial.PARITY_MARK,
            'S': serial.PARITY_SPACE,
        }
        self.ser = serial.Serial(
            port=self.port,
            baudrate=self.baudrate,
            bytesize=self.bytesize,
            parity=parity_map.get(self.parity, serial.PARITY_NONE),
            stopbits=self.stopbits,
            timeout=self.timeout_s,
        )
        # Clean input buffer
        try:
            self.ser.reset_input_buffer()
            self.ser.reset_output_buffer()
        except Exception:
            pass

    def close(self):
        if self.ser:
            try:
                self.ser.close()
            finally:
                self.ser = None

    def is_open(self) -> bool:
        return bool(self.ser and self.ser.is_open)

    @staticmethod
    def _crc16(data: bytes) -> int:
        crc = 0xFFFF
        for b in data:
            crc ^= b
            for _ in range(8):
                if (crc & 0x0001) != 0:
                    crc >>= 1
                    crc ^= 0xA001
                else:
                    crc >>= 1
        return crc & 0xFFFF

    def _write(self, data: bytes):
        assert self.ser is not None
        self.ser.write(data)
        self.ser.flush()

    def _read_exactly(self, n: int, overall_timeout: float) -> bytes:
        assert self.ser is not None
        deadline = time.monotonic() + overall_timeout
        buf = bytearray()
        while len(buf) < n:
            chunk = self.ser.read(n - len(buf))
            if chunk:
                buf.extend(chunk)
                continue
            if time.monotonic() > deadline:
                break
        if len(buf) != n:
            raise TimeoutError(f'timeout reading {n} bytes (got {len(buf)})')
        return bytes(buf)

    def read_holding_registers(self, slave_id: int, start_address: int, quantity: int) -> List[int]:
        if slave_id < 1 or slave_id > 247:
            raise ModbusError('invalid slave_id')
        if quantity < 1 or quantity > 125:
            raise ModbusError('invalid quantity')
        if start_address < 0 or start_address > 0xFFFF:
            raise ModbusError('invalid start_address')

        # Build request: [slave][0x03][addr_hi][addr_lo][qty_hi][qty_lo][crc_lo][crc_hi]
        pdu = bytes([
            0x03,
            (start_address >> 8) & 0xFF,
            start_address & 0xFF,
            (quantity >> 8) & 0xFF,
            quantity & 0xFF,
        ])
        adu_wo_crc = bytes([slave_id]) + pdu
        crc = self._crc16(adu_wo_crc)
        req = adu_wo_crc + bytes([crc & 0xFF, (crc >> 8) & 0xFF])

        self._write(req)

        # Response handling
        # Read address + function first
        header = self._read_exactly(2, self.timeout_s)
        resp_slave = header[0]
        func = header[1]
        if resp_slave != slave_id:
            raise ModbusError(f'unexpected slave id {resp_slave}')

        if func & 0x80:
            # Exception response: next byte is exception code, followed by CRC
            ex_code = self._read_exactly(1, self.timeout_s)[0]
            trailer = self._read_exactly(2, self.timeout_s)
            frame_wo_crc = header + bytes([ex_code])
            crc_calc = self._crc16(frame_wo_crc)
            crc_recv = trailer[0] | (trailer[1] << 8)
            if crc_calc != crc_recv:
                raise ModbusError('CRC mismatch on exception response')
            raise ModbusError(f'Modbus exception code {ex_code}')

        # Normal response: next byte is byte count
        byte_count = self._read_exactly(1, self.timeout_s)[0]
        data = self._read_exactly(byte_count, self.timeout_s)
        crc_bytes = self._read_exactly(2, self.timeout_s)
        frame_wo_crc = header + bytes([byte_count]) + data
        crc_calc = self._crc16(frame_wo_crc)
        crc_recv = crc_bytes[0] | (crc_bytes[1] << 8)
        if crc_calc != crc_recv:
            raise ModbusError('CRC mismatch')

        if byte_count != quantity * 2:
            raise ModbusError(f'unexpected byte count {byte_count}, expected {quantity*2}')

        regs: List[int] = []
        for i in range(0, byte_count, 2):
            regs.append((data[i] << 8) | data[i + 1])
        return regs

File: iot_driver_copilot/industrialtest_3/modbus.go
package main

import (
	"errors"
	"fmt"
	"time"

	serial "go.bug.st/serial"
)

func modbusCRC16(data []byte) uint16 {
	var crc uint16 = 0xFFFF
	for _, b := range data {
		crc ^= uint16(b)
		for i := 0; i < 8; i++ {
			if (crc & 0x0001) != 0 {
				crc = (crc >> 1) ^ 0xA001
			} else {
				crc >>= 1
			}
		}
	}
	return crc
}

func buildReadHoldingRegistersRequest(slaveID byte, startAddr uint16, qty uint16) []byte {
	req := []byte{
		slaveID,
		0x03,
		byte(startAddr >> 8), byte(startAddr & 0xFF),
		byte(qty >> 8), byte(qty & 0xFF),
	}
	crc := modbusCRC16(req)
	req = append(req, byte(crc&0xFF), byte(crc>>8)) // CRC Low, CRC High
	return req
}

func readExact(port serial.Port, expected int, timeout time.Duration) ([]byte, error) {
	_ = port.SetReadTimeout(timeout)
	buf := make([]byte, expected)
	total := 0
	deadline := time.Now().Add(timeout)
	for total < expected {
		if time.Now().After(deadline) {
			return nil, fmt.Errorf("timeout waiting for %d bytes", expected)
		}
		n, err := port.Read(buf[total:])
		if err != nil {
			return nil, err
		}
		if n == 0 {
			// small sleep to avoid busy loop if device is slow
			time.Sleep(5 * time.Millisecond)
			continue
		}
		total += n
	}
	return buf, nil
}

func readHoldingRegisters(port serial.Port, slaveID byte, startAddr uint16, qty uint16, timeout time.Duration) ([]byte, error) {
	req := buildReadHoldingRegistersRequest(slaveID, startAddr, qty)
	_, err := port.Write(req)
	if err != nil {
		return nil, err
	}
	expected := 5 + int(2*qty) // addr(1)+func(1)+byteCount(1)+data(2*qty)+crc(2)
	resp, err := readExact(port, expected, timeout)
	if err != nil {
		return nil, err
	}
	if len(resp) < 5 {
		return nil, errors.New("short response")
	}
	if resp[0] != slaveID {
		return nil, fmt.Errorf("unexpected slave id: got %d", resp[0])
	}
	if resp[1] != 0x03 {
		return nil, fmt.Errorf("unexpected function code: 0x%X", resp[1])
	}
	byteCount := int(resp[2])
	if byteCount != int(2*qty) {
		return nil, fmt.Errorf("unexpected byte count: %d", byteCount)
	}
	// Verify CRC
	crcCalc := modbusCRC16(resp[:len(resp)-2])
	crcRecv := uint16(resp[len(resp)-1])<<8 | uint16(resp[len(resp)-2])
	if crcCalc != crcRecv {
		return nil, fmt.Errorf("crc mismatch: calc 0x%X recv 0x%X", crcCalc, crcRecv)
	}
	return resp, nil
}

func parseRegisterValue(resp []byte, regIndex int, signed bool) int32 {
	// Data starts at resp[3]
	offset := 3 + regIndex*2
	hi := resp[offset]
	lo := resp[offset+1]
	u := uint16(hi)<<8 | uint16(lo)
	if signed {
		return int32(int16(u))
	}
	return int32(u)
}
